constraint-solver/src/boolean_extractor.rs

```diff
@@ -0,0 +1,146 @@
+use std::collections::BTreeMap;
+use std::fmt::Display;
+use std::hash::Hash;
+
+use crate::effect::Effect;
+use crate::quadratic_symbolic_expression::QuadraticSymbolicExpression;
+use crate::quadratic_symbolic_expression::{NoRangeConstraints, RangeConstraintProvider};
+use crate::range_constraint::RangeConstraint;
+use itertools::Itertools;
+use powdr_number::FieldElement;
+
+/// Tries to simplify a quadratic constraint by transforming it into an affine
+/// constraint that makes use of a new boolean variable.
+///
+/// @param constraint The quadratic constraint to transform.
+/// @param var_dispenser A function that returns a new variable that is assumed to be boolean-constrained.
+/// It will only be called if the transformation is performed.
+pub fn extract_boolean<T: FieldElement, V: Ord + Clone + Hash + Eq>(
+    constraint: &QuadraticSymbolicExpression<T, V>,
+    mut var_dispenser: impl FnMut() -> V,
+) -> Option<QuadraticSymbolicExpression<T, V>> {
+    let (left, right) = constraint.try_as_single_product()?;
+    // `constr = 0` is equivalent to `left * right = 0`
+    let offset = left - right;
+    // We only do the transformation if `offset` is known, because
+    // otherwise the constraint stays quadratic.
+    offset.try_to_known()?;
+    // `offset + right = left`
+    // `constr = 0` is equivalent to `right * (right + offset) = 0`
+
+    let z = var_dispenser();
+
+    // We return `right + z * offset == 0`, which is equivalent to the original constraint.
+    Some(right + &(QuadraticSymbolicExpression::from_unknown_variable(z) * offset))
+}
+
+/// Performs the inverse of `extract_boolean` but on a full system of constraints.
+pub fn unextract_boolean<T: FieldElement, V: Ord + Clone + Hash + Eq + Display>(
+    constraints: Vec<QuadraticSymbolicExpression<T, V>>,
+    range_constraints: &impl RangeConstraintProvider<T, V>,
+) -> Vec<QuadraticSymbolicExpression<T, V>> {
+    // TODO does not take into account range constraints from bus interactions.
+
+    // Boolean constraints by variable
+    let mut boolean_constraints = BTreeMap::new();
+    // Other constraints
+    let mut constraints = constraints
+        .into_iter()
+        .filter_map(|constr| {
+            if let Some(var) = is_boolean_constraint(&constr) {
+                boolean_constraints.insert(var.clone(), constr.clone());
+                None
+            } else {
+                Some(constr)
+            }
+        })
+        .collect_vec();
+
+    let constraints_by_variable = constraints
+        .iter()
+        .enumerate()
+        .flat_map(|(i, constr)| {
+            constr
+                .referenced_unknown_variables()
+                .map(move |var| (var.clone(), i))
+        })
+        .into_group_map();
+
+    for (var, constrs) in constraints_by_variable.iter().sorted() {
+        if !boolean_constraints.contains_key(var)
+            && range_constraints.get(var) != RangeConstraint::from_mask(1u64)
```

Comment:
Why can't we just rely on the range constraints? I think we could?

I think this code could be simplified by:
- Removing `boolean_constraints` and keeping `constraints` as is passed to the function.
- In here, continue if there is only one referenced variable, because then it would be the range constraint itself, and the code below would change it to `0`.

I think this would be equivalent and remove some code, including `is_boolean_constraint`.

Reply:
Yeah, saw that! I think we should remove it if it's not used.